Improved Related Posts Logic & Text Alignment

mgiannopoulos24 opened the pull request:
### Summary

This PR updates the related posts logic by replacing Algolia with WordPress's built-in functionality and improves text alignment for better readability.

### Changes
- Removed `axios` import and all Algolia-related code.
- Imported the `wordpress` module.
- Used `wordpress.getPosts(4)` to fetch four random posts.
- Filtered out the current post from recommendations.
- Added a fallback to fetch additional posts if fewer than four remain after filtering.
- Limited the final recommendations to a maximum of four posts.
- Applied **justify alignment** to text for better presentation.

### Closes
Closes #4.

### Screenshots
- Recommendations showcase:
<p align="center">

<img src="https://i.imgur.com/IWvqYGJ.png"/>

</p>

<p align="center">

<img src="https://i.imgur.com/0KRSt2g.png"/>

</p>

- Text before alignment:

<p align="center">

<img src="https://i.imgur.com/8y5TxgH.png"/>

</p>

- Text after alignment:

<p align="center">

<img src="https://i.imgur.com/yrNbhDx.png"/>

</p>

---

Let me know if you need any changes! I also think you should upgrade astro since there's a newer version available.

VangelisDimi commented:
- The `wordpress.getPosts()` method requests the most recent posts.  So now the recommended posts are the same under every posts (the 4 most recent ones).
There is a parameter you can pass in the WordPress API to get random posts which is `orderby=rand`. I haven't managed to make it work though , maybe it doesn't exist in the version we use right now.

- You can use the `exclude` parameter  in the request to exclude the current post ID from the recommendations instead of making a second request to get more posts.

- In the design , the text in the Articles is left aligned so I don't know if we should change it to justified. Although , we should probably remove the space at the start of every paragraph that some articles have.

mgiannopoulos24 commented:
Okay, I will have a look at it and make it right.

VangelisDimi commented:
Maybe also wait for @skourismanolis who is the maintainer to look at it and add any other comments.

skourismanolis commented:
BTW the docs for the wordpress.com api we're using are: [https://developer.wordpress.com/docs/api/](https://developer.wordpress.com/docs/api/)

Tbh i'm not sure how to crack this one, maybe the wp-json api can give us random posts, maybe we need to fetch all the posts of the api at the start of the build and use that for the reccs, maybe memory caching is needed ? I'm not sure, @VangelisDimi is right tho about what he said.

Thanks for taking a crack at this, I'm excited to see what you come up with!

PS: Since this is a large PR and styling changes might be somewhat unrelated, if you want to change the styling of stuff please do so in a separate PR

mgiannopoulos24 commented:
Okay so I have to revert the alignment and use `orderby=rand`.

skourismanolis commented:
sure, but I don't think `orderby=rand` will work as Vangelis mentioned, [it's not documented in the api docs at least](https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/posts/#apidoc-query). Maybe it's part of another API, if you make it work with the other API then sure!

mgiannopoulos24 commented:
I think I fixed it now. 

This modification passes the current post's `id` to the `getPosts` function, which will then use it as the `exclude` parameter in the WordPress API request. This ensures that the current post is not included in the recommended posts. I have also modified the `getPosts` function in `index.js` to accept the `exclude` parameter.

Here is an image that shows posts randomly. 

<p align="center">

<img src="https://i.imgur.com/3FQUDEJ.png"/>

</p>

skourismanolis commented:
Hey! Good work so far, I think you've missed what Vangelis' said in his first comment though regarding the `getPosts()` function, it always returns the **latest** posts, so even with the exclude parameter, we have the almost same posts appearing on every page. You can see this for yourself by opening 3-4 random articles and seeing the results.

That's why I started talking about caching and different APIs, it's not that simple of a feature to build ;) 

Also, you seem to have reformatted the committed files to match your personal prettier configuration, there is `.prettierrc` in the repo, please instruct your formatter to use that since now there are a ton of changes related to that: +6,827 −6,814 😨😨

You can always hit me up on discord if you want guidance/help!

mgiannopoulos24 commented:
Yeah I think I might close the PR and open a new one some other day for the same issue. I messed up a lot of things that I shouldn't like `yarn.lock` and `package.json`.  

Also could you give me your Discord so we can talk some time. Thank you.
